### PR TITLE
DEV: introduces documentation for chat

### PR DESCRIPTION
**Pull request by jjaffeux**
Note this commit also slightly changes internal API: `channel` instead of `getChannel` and `updateCurrentUserChannelNotificationsSettings` instead of `updateCurrentUserChatChannelNotificationsSettings`.

Also `destroyChannel` takes a second param which is the name confirmation instead of an optional object containing this confirmation. This is to enforce the fact that it's required.

In the future a top level jsdoc config file could be used instead of the hack tempfile, but while it's only an experiment for chat, it's probably good enough.

### CODE REVIEW

File: plugins/chat/lib/tasks/chat_doc.rake [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+task "chat:doc" do
+  unless system("command -v jsdoc2md >/dev/null;")
```
Comment (martin-brennan) [inline plugins/chat/lib/tasks/chat_doc.rake:4]: This is cool, I didn't know there was a tool that could do jsdoc into markdown

**Review by martin-brennan — COMMENTED** [review]
I like the implementation and the idea behind it, personally I am not sure how often I would go into the generated markdown docs to read them rather than just reading the JSDoc inline in the code. The nicest thing is the method directory at the top IMO:

image

Also will autolinking of classes work too, e.g. `Collection` linking to the `Collection` class docs from `ChatApi`?

I guess my main question is who the target audience for the generated markdown docs are? I am not trying to be too pessimistic, I definitely think the jsDoc additions are great. Usually the way I work is grepping through the code and looking at docs inline, usually I only look at docs for external things e.g. the ruby docs, MDN, capybara DSL etc.

**Comment by jjaffeux** [comment]
I generally agree with the question and I thinks it's hard to answer this when you have been coding it yourself. ofc documentation doesnt seem super useful when you know the code. One audience I can think of is people joining chat project and learning about it through documentation.

> usually I only look at docs for external things e.g. the ruby docs, MDN, capybara DSL etc.

Mostly because we dont have documentation IMO.

Generating the doc is almost no work so as long as jsdoc is useful Im fine with it, we will see over time if it's useful or not, but it doesnt take me time to do.

Also note that documenting code also helps having better code, as you can see in the PR, I changed few methods while documenting it as it wasnt very clear or has been showing some inconsistencies.

**Comment by martin-brennan** [comment]
@jjaffeux yes for sure don't get me wrong I agree that documenting the code in general and adding JSdoc is super useful, I just don't personally think I would find the markdown docs that are generated out of it that useful (since I can just read the JSdoc in the code).

**Comment by jjaffeux** [comment]
> @jjaffeux yes for sure don't get me wrong I agree that documenting the code in general and adding JSdoc is super useful, I just don't personally think I would find the markdown docs that are generated out of it that useful (since I can just read the JSdoc in the code).

yes me too, not sure, I use this a lot for external projects. So maybe I might end up using it if it grows. As long as it doesnt take much time/energy I think Im fine with doing it.

**Review by martin-brennan — APPROVED** [review]

File: plugins/chat/assets/javascripts/discourse/services/chat-api.js [hunk 2]
```diff
@@ -154,20 +234,50 @@ export default class ChatApi extends Service {
     });
   }
 
+  /**
+   * Creates a channel archive.
+   * @param {number} channelId - The ID of the channel.
+   * @param {object} data - Params of the archive.
+   * @param {string} data.selection - "new_topic" or "existing_topic".
+   * @param {string} [data.title] - Title of the topic when creating a new topic.
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/services/chat-api.js:242]: brackets mean it's optional

**Comment by jjaffeux** [comment]
> I like the implementation and the idea behind it, personally I am not sure how often I would go into the generated markdown docs to read them rather than just reading the JSDoc inline in the code. The nicest thing is the method directory at the top IMO:
> 
> image
> 
> Also will autolinking of classes work too, e.g. `Collection` linking to the `Collection` class docs from `ChatApi`?
> 
> I guess my main question is who the target audience for the generated markdown docs are? I am not trying to be too pessimistic, I definitely think the jsDoc additions are great. Usually the way I work is grepping through the code and looking at docs inline, usually I only look at docs for external things e.g. the ruby docs, MDN, capybara DSL etc.

Something you might not know as a VIM user, but vscode will show you the doc while you type:

Screenshot 2023-01-11 at 17 46 29

File: .github/workflows/documentation.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: Yarn install
+        run: yarn install
+
+      - name: jsdoc-to-markdown install
+        run: |
+          yarn global add jsdoc-to-markdown
+          yarn global upgrade jsdoc-to-markdown
```
Comment (davidtaylorhq) [inline .github/workflows/documentation.yml:66]: We should be able to remove this now?

File: plugins/chat/lib/tasks/chat_doc.rake [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+task "chat:doc" do
+  ["plugins/chat/assets/javascripts/discourse/services/chat-api.js"].each do |file|
+    origin = File.join(Rails.root, file)
+    filename = File.basename(file, ".*")
+    destination = File.join(Rails.root, "plugins/chat/docs/#{filename}.md")
+    config = File.join(Rails.root, ".jsdoc")
+
+    `jsdoc2md -c #{config} #{origin} > #{destination}`
```
Comment (davidtaylorhq) [inline plugins/chat/lib/tasks/chat_doc.rake:10]: I expect this is still using the globally-installed version. `yarn jsdoc2md ...` should do it

File: .github/workflows/documentation.yml (hunk 3 above)
Comment (jjaffeux) [reply .github/workflows/documentation.yml]: Ah I was doing and didn't push it because I got sidetracked by another issue, but yes :+1:

**Review by davidtaylorhq — APPROVED** [review]

**Comment by martin-brennan** [comment]
> Something you might not know as a VIM user, but vscode will show you the doc while you type:

Heh yes us vim dinosaurs do occasionally have nice things like this these days :P 

image

**Pull request merged**